examples/src/main/java/io/grpc/examples/healthservice/HealthServiceServer.java

```diff
@@ -0,0 +1,149 @@
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+/**
+ * Server that manages startup/shutdown of a {@code Greeter} server.
+ */
+public class HealthServiceServer {
+  private static final Logger logger = Logger.getLogger(HealthServiceServer.class.getName());
+
+  private Server server;
+  private HealthStatusManager health;
+
+  private void start() throws IOException {
+    /* The port on which the server should run */
+    int port = 50051;
+    health = new HealthStatusManager();
+    server = Grpc.newServerBuilderForPort(port, InsecureServerCredentials.create())
+        .addService(new GreeterImpl())
+        .addService(health.getHealthService())
+        .build()
+        .start();
+    logger.info("Server started, listening on " + port);
+    Runtime.getRuntime().addShutdownHook(new Thread() {
+      @Override
+      public void run() {
+        // Use stderr here since the logger may have been reset by its JVM shutdown hook.
+        System.err.println("*** shutting down gRPC server since JVM is shutting down");
+        try {
+          HealthServiceServer.this.stop();
+        } catch (InterruptedException e) {
+          e.printStackTrace(System.err);
+        }
+        System.err.println("*** server shut down");
+      }
+    });
+
+    health.setStatus("", ServingStatus.SERVING);
+  }
+
+  private void stop() throws InterruptedException {
+    if (server != null) {
+      server.shutdown().awaitTermination(30, TimeUnit.SECONDS);
+    }
+  }
+
+  /**
+   * Await termination on the main thread since the grpc library uses daemon threads.
+   */
+  private void blockUntilShutdown() throws InterruptedException {
+    if (server != null) {
+      server.awaitTermination();
+    }
+  }
+
+  /**
+   * Main launches the server from the command line.
+   */
+  public static void main(String[] args) throws IOException, InterruptedException {
+    System.setProperty("java.util.logging.SimpleFormatter.format",
+        "%1$tH:%1$tM:%1$tS %4$s %2$s: %5$s%6$s%n");
+
+    final HealthServiceServer server = new HealthServiceServer();
+    server.start();
+    server.blockUntilShutdown();
+  }
+
+  private class GreeterImpl extends GreeterGrpc.GreeterImplBase {
+    boolean isServing = true;
+
+    @Override
+    public void sayHello(HelloRequest req, StreamObserver<HelloReply> responseObserver) {
+      if (!isServing) {
+        responseObserver.onError(
+            Status.INTERNAL.withDescription("Not Serving right now").asRuntimeException());
+        return;
+      }
+
+      if (validateRequest(req)) {
+        HelloReply reply = HelloReply.newBuilder().setMessage("Hello " + req.getName()).build();
+        responseObserver.onNext(reply);
+        responseObserver.onCompleted();
+      } else {
+        responseObserver.onError(
+            Status.INVALID_ARGUMENT.withDescription("Offended by short name").asRuntimeException());
+      }
+    }
+
+    private boolean validateRequest(HelloRequest req) {
```

Comment:
nit: Name it isNameLongEnough()?

Reply:
ok.  Moved state changing logic out of method since "is" methods shouldn't have side effects.